### PR TITLE
OSDOCS-1048 Improve Ingress sharding documentation

### PR DESCRIPTION
**Pull request by bmcelvee**
https://issues.redhat.com/browse/OSDOCS-1048

### CODE REVIEW

**Comment by bmcelvee** [comment]
Preview docs:
https://osdocs-1048-router-sharding--ocpdocs.netlify.app/openshift-enterprise/latest/networking/ingress-operator.html#nw-ingress-sharding_configuring-ingress

https://osdocs-1048-router-sharding--ocpdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/routing-optimization.html

**Comment by bmcelvee** [comment]
As far as placement in the docs, I have several related PRs in progress. So ultimately, the sharding-specific content may get pulled into it's own assembly.

File: scalability_and_performance/routing-optimization.adoc [hunk 1]
```diff
@@ -9,4 +9,6 @@ The {product-title} HAProxy router scales to optimize performance.
 
 include::modules/baseline-router-performance.adoc[leveloffset=+1]
 
+For more information on Ingress sharding, see xref:../networking/ingress-operator.adoc#nw-ingress-sharding-route-labels[Configuring Ingress Controller sharding by using route labels] and xref:../networking/ingress-operator.adoc#nw-ingress-sharding-namespace-labels[Configuring Ingress Controller sharding by using namespace labels].
+
```
Comment (bmcelvee) [inline scalability_and_performance/routing-optimization.adoc:13]: Links may change.

**Comment by bmcelvee** [comment]
@danehans, @Miciah, @frobware - when you get a chance, would one of you mind taking a look at this PR please?

In the request there are two major points that I'm missing:
1. Limitations of sharding - I'm not certain where to find this information. Do you know where I could find more information on the limitations? If so, could you point me to it?
2. Examples - I will try and come up with these

Thanks!

**Comment by bmcelvee** [comment]
@lihongan would you mind reviewing this PR please? There are no new procedures, just clarifying our documentation on Ingress sharding. Thank you!

**Comment by lihongan** [comment]
The changes look good to me. Thank you @bmcelvee

**Comment by bmcelvee** [comment]
@openshift/team-documentation please peer review, thanks!

File: modules/baseline-router-performance.adoc [hunk 2]
```diff
@@ -1,12 +1,9 @@
 [id="baseline-router-performance_{context}"]
-= Baseline router performance
+= Baseline Ingress Controller (router) performance
 
-The {product-title} router is the Ingress point for all external traffic
-destined for {product-title} services.
+The {product-title} Ingress Controller, or router, is the Ingress point for all external traffic destined for {product-title} services.
```
Comment (bobfuru) [inline modules/baseline-router-performance.adoc:4]: Should we _always_ capitalize "Ingress"? With "Ingress point", it seems like a common/general usage and I would think of lowercasing it.

File: modules/baseline-router-performance.adoc [hunk 3]
```diff
@@ -1,12 +1,9 @@
...
 
-When evaluating a single HAProxy router performance in terms of
-HTTP requests handled per second, the performance varies depending
-on many factors. In particular:
+When evaluating a single HAProxy router performance in terms of HTTP requests handled per second, the performance varies depending on many factors. In particular:
 
 * HTTP keep-alive/close mode
```
Comment (bobfuru) [inline modules/baseline-router-performance.adoc:8]: Not in scope for this PR, but thinking it would be nice to first-cap in the list, per ISG (p. 20): "Capitalize the first word of each item in a vertical list."

File: modules/baseline-router-performance.adoc [hunk 4]
```diff
@@ -22,11 +19,7 @@ on many factors. In particular:
 
 * underlying infrastructure (network/SDN solution, CPU, and so on)
 
-While performance in your specific environment will vary, Red Hat lab
-tests on a public cloud instance of size 4 vCPU/16GB RAM, a single
-HAProxy router handling 100 routes terminated by backends serving
-1kB static pages is able to handle the following number of transactions
-per second.
+While performance in your specific environment will vary, Red Hat lab tests on a public cloud instance of size 4 vCPU/16GB RAM, a single HAProxy router handling 100 routes terminated by backends serving 1kB static pages is able to handle the following number of transactions per second.
```
Comment (bobfuru) [inline modules/baseline-router-performance.adoc:22]: Red Hat _Lab_? Haven't come across this before.
Also, something is off in this sentence. I think you might be able to start a new sentence at, "A single HAProxy...."

File: modules/baseline-router-performance.adoc [hunk 5]
```diff
@@ -50,17 +43,11 @@ In HTTP close (no keep-alive) scenarios:
 |re-encrypt |2320|2941
 |===
 
-Default router configuration with `ROUTER_THREADS=4` was used and two
+Default Ingress Controller configuration with `ROUTER_THREADS=4` was used and two
 different endpoint publishing strategies (LoadBalancerService/HostNetwork) tested.
```
Comment (bobfuru) [inline modules/baseline-router-performance.adoc:47]: s/tested/were tested/

File: modules/baseline-router-performance.adoc [hunk 6]
```diff
@@ -50,17 +43,11 @@ In HTTP close (no keep-alive) scenarios:
...
-keep-alive, a single HAProxy router is capable of saturating 1 Gbit
-NIC at page sizes as small as 8 kB.
+TLS session resumption was used for encrypted routes. With HTTP keep-alive, a single HAProxy router is capable of saturating 1 Gbit NIC at page sizes as small as 8 kB.
 
-When running on bare metal with modern processors, you can expect roughly
-twice the performance of the public cloud instance above. This
-overhead is introduced by the virtualization layer in place on public clouds and
-holds mostly true for private cloud-based virtualization as well. The following
-table is a guide on how many applications to use behind the router:
+When running on bare metal with modern processors, you can expect roughly twice the performance of the public cloud instance above. This overhead is introduced by the virtualization layer in place on public clouds and holds mostly true for private cloud-based virtualization as well. The following table is a guide on how many applications to use behind the router:
```
Comment (bobfuru) [inline modules/baseline-router-performance.adoc:50]: "guide on how" sounds odd to me, maybe "guide to how"?

File: modules/nw-ingress-sharding.adoc [hunk 7]
```diff
@@ -0,0 +1,20 @@
+// Module included in the following assemblies:
+//
+// * configuring_ingress_cluster_traffic/configuring-ingress-cluster-traffic-ingress-controller.adoc
+// * ingress-operator.adoc
+
+[id="nw-ingress-sharding_{context}"]
+= Ingress Controller sharding
+
+As the primary mechanism for traffic to enter the cluster the demands on the Ingress Controller, or router, can be significant. As cluster administrator, you can shard the routes to:
```
Comment (bobfuru) [inline modules/nw-ingress-sharding.adoc:9]: "to enter the cluster the demands" - something is off here, I think it needs a comma:

> As the primary mechanism for traffic to enter the cluster, the demands on the Ingress Controller

Comment (bobfuru) [inline modules/nw-ingress-sharding.adoc:9]: s/As cluster administrator/As a cluster administrator/

File: modules/nw-ingress-sharding.adoc [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+* Balance Ingress Controllers, or routers, with several routes to speed up responses to changes.
+* Allocate certain routes to have different reliability guarantees than other routes.
+* Allow certain Ingress Controllers to have different policies defined.
+* Allow only specific routes to use additional features.
+* Expose different routes on different addresses, which for example, would allow external and internal users see different routes.
```
Comment (bobfuru) [inline modules/nw-ingress-sharding.adoc:15]: I would reword this slightly; "which for example" is an odd phrase, and I'd also swap external/internal:

> Expose different routes on different addresses so that internal and external users can see different routes, for example.

File: modules/nw-ingress-sharding.adoc [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+There are different methods of Ingress Controller sharding:
+
+* Ingress Controller sharding by using route labels
+* Ingress Controller sharding by using namespace labels
```
Comment (bobfuru) [inline modules/nw-ingress-sharding.adoc:20]: Can you shorten this to just:

> Ingress Controller can use either route labels or namespace labels as a sharding method.

File: scalability_and_performance/routing-optimization.adoc [hunk 10]
```diff
@@ -9,4 +9,6 @@ The {product-title} HAProxy router scales to optimize performance.
 
 include::modules/baseline-router-performance.adoc[leveloffset=+1]
 
+For more information on Ingress sharding, see xref:../networking/ingress-operator.adoc#nw-ingress-sharding-route-labels[Configuring Ingress Controller sharding by using route labels] and xref:../networking/ingress-operator.adoc#nw-ingress-sharding-namespace-labels[Configuring Ingress Controller sharding by using namespace labels].
```
Comment (bobfuru) [inline scalability_and_performance/routing-optimization.adoc:12]: The xrefs are landing at the top of the page for me. Perhaps the tags need `_configuring-ingress` on the end? 
`xref:../networking/ingress-operator.adoc#nw-ingress-sharding-route-labels_configuring-ingress`
`xref:../networking/ingress-operator.adoc#nw-ingress-sharding-namespace-labels_configuring-ingress`

**Review by bobfuru — COMMENTED** [review]
Overall, this is great and LGTM! The descriptions are really well written and easy to follow. Kind of makes me miss writing performance/optimization docs. :) Just a few suggestions for rewording and a couple of links not working properly.

File: modules/baseline-router-performance.adoc (hunk 2 above)
Comment (bmcelvee) [reply modules/baseline-router-performance.adoc:4]: Yeah...since it's a Kubernetes resource and they capitalize it, I've been following that pattern. It's definitely something we can look into in the future (especially since it looks like they're now not capitalizing controller in Ingress controller).

File: modules/baseline-router-performance.adoc (hunk 3 above)
Comment (bmcelvee) [reply modules/baseline-router-performance.adoc:8]: Good idea and easy fix!

File: modules/baseline-router-performance.adoc (hunk 4 above)
Comment (bmcelvee) [reply modules/baseline-router-performance.adoc:22]: This content is from the Scalability and Performance team so they do a lot of testing--probably where the labs portion comes from. :)

File: modules/nw-ingress-sharding.adoc (hunk 8 above)
Comment (bmcelvee) [reply modules/nw-ingress-sharding.adoc:15]: Thank you! That one was bothering me too.
